ui/src/dashboards/components/QueryEditor.tsx

```diff
@@ -0,0 +1,465 @@
+  onUpdate: (text: string) => void
+  config: QueryConfig
+  isInDataExplorer?: boolean
+  templates: Template[]
+}
+
+const CODE_MIRROR_OPTIONS = {
+  tabIndex: 1,
+  mode: 'influxQL',
+  readonly: false,
+  lineNumbers: false,
+  autoRefresh: true,
+  theme: 'influxql',
+  completeSingle: false,
+  lineWrapping: true,
+}
+
+const TEMPLATE_START = ':'.length
+
+const NOOP = () => {}
+const NULL_RESOLUTION = null
+const SHOW_TEMPLATES = 'Show Raw Query'
+const HIDE_TEMPLATES = 'Hide Raw Query'
+
+const BLURRED_EDITOR_STATE = {
+  focused: false,
+  isTemplating: false,
+  isShowingTemplateValues: false,
+}
+
+@ErrorHandling
+class QueryEditor extends Component<Props, State> {
+  public static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+    if (prevState.editorValue !== nextProps.query) {
+      return {
+        ...BLURRED_EDITOR_STATE,
+        selectedTemplate: {
+          tempVar: _.get(nextProps.templates, ['0', 'tempVar'], ''),
+        },
+        filteredTemplates: nextProps.templates,
+        editorValue: nextProps.query,
+        queryText: nextProps.query,
+      }
+    }
+
+    return null
+  }
+
+  private editor?: IInstance
+
+  constructor(props: Props) {
+    super(props)
+
+    this.editor = null
+
+    this.state = {
+      ...BLURRED_EDITOR_STATE,
+      selectedTemplate: {
+        tempVar: _.get(props.templates, ['0', 'tempVar'], ''),
+      },
+      filteredTemplates: props.templates,
+      editorValue: props.query,
+      queryText: props.query,
+    }
+  }
+
+  public render() {
+    const {
+      config: {status},
+    } = this.props
+
+    const {
+      editorValue,
+      isTemplating,
+      selectedTemplate,
+      filteredTemplates,
+      isShowingTemplateValues,
+      focused,
+    } = this.state
+
+    const options = {
+      ...CODE_MIRROR_OPTIONS,
+      readOnly: isShowingTemplateValues,
+    }
+
+    return (
+      <div className="query-editor">
+        <div className={this.queryCodeClassName}>
+          <ReactCodeMirror
+            autoFocus={true}
+            autoCursor={true}
+            options={options}
+            onTouchStart={NOOP}
+            value={editorValue}
+            onChange={this.handleChange}
+            onBlur={this.handleBlurEditor}
+            onFocus={this.handleFocusEditor}
+            onKeyDown={this.handleKeyDownEditor}
+            onDblClick={this.handleUnmountEditor}
+            editorDidMount={this.handleMountEditor}
+            onBeforeChange={this.handleUpdateEditorValue}
+          />
+        </div>
+        <div
+          className={classnames('varmoji', {
+            'varmoji-rotated': isTemplating,
+            focus: focused,
+          })}
+        >
+          <div className="varmoji-container">
+            <div className="varmoji-front">
+              <QueryStatus status={status}>
+                <button
+                  onMouseDown={this.handleToggleFocus}
+                  onClick={this.handleToggleTemplateValues}
```

Comment:
i suppose this could be called something like `handleToggleShowRawQuery`, but just thinking out loud